### ISSUE TITLE
RFC:  adding O() notation in the documentation for functions, like Haskell

### ISSUE THREAD

**Issue by ScottPJones** [JuliaLang/julia#11161]
I noticed that the Haskell string functions actually have O(1) or O(n) as the very first thing in the documentation of many of their functions...
What do people think about adding that to the Julia documentation?
Julians in general do seem to be rather concerned about performance, that might be useful information...

**Comment by pao** [comment]
Also true for the C++ STL. Seems reasonable to document.

**Comment by StefanKarpinski** [comment]
+1 – excellent idea. Applies more generally than just string functions, but it's a good place to start.

**Comment by garrison** [comment]
I'd also suggest that it document the worst case performance, not the average case, in places where these differ.

**Comment by ScottPJones** [comment]
Good point, @garrison.  The Haskell docs I saw weren't that complete... they only had the simple ones, i.e. O(1) or O(n) for the string ops.

**Comment by stevengj** [comment]
Should we use Θ rather than O, in order to give more information than a mere upper bound?  Or is that too pedantic, and the colloquial use of O for Θ is good enough?

**Comment by ScottPJones** [comment]
@stevengj I think most people (even some of us who went to MIT) would not know what that meant, so I'd vote for simple O(), and for many of the string cases, it really is that (O(1) or O(n)), not an upper bound.  Don't confuse us poor people who aren't as smart as all of you guys! :grinning:

**Comment by carnaval** [comment]
(fat fingers sorry)

I'm all for technical correctness but I think that people who already know the difference are used to the popular meaning, and people who don't will just be confused.

**Comment by ScottPJones** [comment]
@carnaval and here I was just beginning to think that that is the standard response to any issue I raise! :grinning:  close it down as fast as possible to shut me up!

**Comment by garrison** [comment]
I would be supportive of Θ.  It looks enough like an O that people who are confused should still know what it means.

**Comment by ScottPJones** [comment]
@garrison I still don't know what it means! :grinning:

**Comment by StefanKarpinski** [comment]
+1 for using Θ and being technically accurate.

**Comment by ScottPJones** [comment]
Please explain for those of us who are technically challenged exactly what the difference is, so that when I add documentation, I can do it correctly (I did study this stuff under Ron Rivest, once upon a time, but my memory is not all that good any more!)

**Comment by StefanKarpinski** [comment]
Wikipedia has a pretty good summary here:

http://en.wikipedia.org/wiki/Big_O_notation#Family_of_Bachmann.E2.80.93Landau_notations

Basically, O(n) include constant functions, logarithmic, etc., whereas Θ(n) only includes functions that are not only linearly bounded above but also below. This seems to usually be what people mean.

**Comment by ScottPJones** [comment]
So, `O(1)` stays `O(1)`, and `O(n)` [where you walk over each character of a string] would also stay that way?  I'd read the definition of Θ on Wikipedia already, I just wasn't clear on the difference between the two...

**Comment by garrison** [comment]
@ScottPJones see also [StackOverflow](http://stackoverflow.com/questions/471199/what-is-the-difference-between-%CE%98n-and-on).

**Comment by ScottPJones** [comment]
@garrison Thanks, that link was exactly the ticket!  So, I really need to use Θ(g(n)), or Ω(g(n)) and  O(g(n)), to be technically accurate, correct?  (I can see cases with Ω(1) and O(n), for example... say if I'm caching the length for long strings that aren't `DirectIndexString`... right?)

**Comment by ScottPJones** [comment]
I also think if we use theta instead of, or with O, in the documentation, it will deserve a note in the documentation, for those who *might* be confused...

**Comment by Mike43110** [comment]
+1 for being accurate

**Comment by StefanKarpinski** [comment]
Well, O(f(n)) notation is also correct, just not as precise as Θ(f(n)).

**Comment by IainNZ** [comment]
+1 to this, if there weren't enough already. Its a good way to reduce what I'd call "computational complexity anxiety"

**Comment by toivoh** [comment]
@ScottPJones: Agree that theta deserves a note somewhere (a link to the Wikipedia page?).

**Comment by ScottPJones** [comment]
@toivoh I personally found the StackOverflow link that @garrison pointed out much more helpful than the Wikipedia page, so maybe point to both, after a nice concise description in the Julia docs?

**Comment by Moelf** [comment]
to make this `Hacktoberfest`-able, we should prepare a list of functions that people are most likely interested knowing `O()` of

**Comment by udohjeremiah** [comment]
I love this discussion and I'd love to be a part of it, maybe in contributing to the functions people are likely interested knowing `0()` of as @Moelf pointed out. So where is this issue currently at so I may follow up?

**Comment by LilithHafner** [comment]
If we're really concerned about being technically accurate, we should use O(f(n)), not Θ(f(n)) because if we use `Θ` notation and then make an asymptotic performance improvement it's technically a breaking change.

---

A way to help would be to compile a list of candidate functions and post them here.

**Comment by stevengj** [comment]
A difficulty here is that, technically, we can only be sure of the complexity of particular *methods*, not functions.  A package is always free to override a function in a way that has worse complexity (even if it avoids type piracy).

This is different from Haskell or C++ STL, where there is really only a single polymorphic implementation.
